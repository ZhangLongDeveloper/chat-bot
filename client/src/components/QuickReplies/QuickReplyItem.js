import React, { Component } from 'react';
import PropTypes from 'prop-types';

class QuickReplyItem extends Component {
  handleClick = () => {
    const { onClick, data: { type, value } } = this.props;

    if (onClick) {
      onClick(type, value);
    }
  };

  renderLink() {
    const { text, value } = this.props.data;

    return (
      <a
        href={value}
        rel="noopener noreferrer"
        target="_blank"
        className="waves-effect waves-light btn light-blue hoverable"
      >
        {text}
      </a>
    );
  }

  renderButton() {
    return (
      <button
        disabled={this.props.disabled}
        onClick={this.handleClick}
        className="waves-effect waves-light btn light-blue hoverable"
      >
        {this.props.data.text}
      </button>
    );
  }

  renderItem() {
    switch (this.props.data.type) {
      case 'link':
        return this.renderLink();
      case 'text':
      case 'trigger':
        return this.renderButton();
      default:
        return null;
    }
  }

  render() {
    return (
      <div className="quick-reply-containter">
        { this.renderItem() }
      </div>
    );
  }
}

QuickReplyItem.propTypes = {
  data: PropTypes.object.isRequired,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
};

QuickReplyItem.defaultProps = {
  onClick: null,
  disabled: false,
};

export default QuickReplyItem;
