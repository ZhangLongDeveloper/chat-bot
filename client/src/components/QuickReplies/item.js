import React from 'react';
import PropTypes from 'prop-types';

const QuickReply = ({ data, onClick, disabled }) => {
  const { type, text, value } = data;

  const handleClick = () => {
    onClick(type, value);
  };

  const renderLink = () => (
    <a href={value} rel="noopener noreferrer" target="_blank" className="waves-effect waves-light btn light-blue hoverable">
      {text}
    </a>
  );

  const renderButton = () => (
    <button disabled={disabled} onClick={handleClick} className="waves-effect waves-light btn light-blue hoverable">
      {text}
    </button>
  );

  const renderItem = () => {
    switch (type) {
      case 'link':
        return renderLink();
      case 'text':
      case 'trigger':
        return renderButton();
      default:
        return null;
    }
  };

  return (
    <div className="quick-reply-containter">
      { renderItem() }
    </div>
  );
};

QuickReply.propTypes = {
  data: PropTypes.object.isRequired,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
};

QuickReply.defaultProps = {
  onClick: () => {},
  disabled: false,
};

export default QuickReply;
