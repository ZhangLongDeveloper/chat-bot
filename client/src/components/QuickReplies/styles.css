.quick-reply-containter {
  display: inline-block;
  margin: 5px;
}