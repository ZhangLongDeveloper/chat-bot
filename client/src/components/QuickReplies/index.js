import React, { Component } from 'react';
import PropTypes from 'prop-types';

import QuickRely from './item';

import './styles.css';

class QuickReplies extends Component {
  state = {
    answered: false,
  };

  handleReply = (type, value) => {
    this.props.onReply(type, value);
    this.setState({ answered: true });
  }

  render() {
    const { data } = this.props;
    const { answered } = this.state;

    if (data && data.values && data.values.length) {
      return (
        <div className="quick-replies-container">
          <p>{data.message}</p>
          {data.values.map(item => (
            <QuickRely data={item} key={item.id} onClick={this.handleReply} disabled={answered} />
          ))}
        </div>
      );
    }

    return <span className="red-text">Oops, something went wrong... :(</span>;
  }
}

QuickReplies.propTypes = {
  data: PropTypes.object.isRequired,
  onReply: PropTypes.func.isRequired,
};

export default QuickReplies;
