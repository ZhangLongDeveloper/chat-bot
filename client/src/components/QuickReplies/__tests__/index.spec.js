import React from 'react';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';

import QuickReplies from '../../QuickReplies';

jest.mock('../Item', () => 'QuickRely');

const mockComponent = props => (
  <QuickReplies {...props} />
);

describe('components/<QuickReplies />', () => {
  it('should render component with no data', () => {
    const props = { data: {} };
    const tree = create(mockComponent(props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with data', () => {
    const props = {
      data: {
        values: [
          { id: 'id-1' },
          { id: 'id-2' },
          { id: 'id-3' },
        ],
      },
    };
    const tree = create(mockComponent(props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call onReply and setState when handleReply is called', () => {
    const onReply = jest.fn();
    const setState = jest.fn();
    const type = 'type';
    const value = 'value';
    const props = { onReply };

    const instance = mount(mockComponent(props)).instance();

    instance.setState = setState;
    instance.handleReply(type, value);

    expect(onReply).toHaveBeenCalledWith(type, value);
    expect(setState).toHaveBeenCalledWith({ answered: true });
  });
});
