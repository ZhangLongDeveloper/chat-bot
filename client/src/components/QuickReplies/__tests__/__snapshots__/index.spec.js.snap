// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/<QuickReplies /> should render component with data 1`] = `
<div
  className="quick-replies-container"
>
  <p />
  <QuickReplyItem
    data={
      Object {
        "id": "id-1",
      }
    }
    disabled={false}
    onClick={[Function]}
  />
  <QuickReplyItem
    data={
      Object {
        "id": "id-2",
      }
    }
    disabled={false}
    onClick={[Function]}
  />
  <QuickReplyItem
    data={
      Object {
        "id": "id-3",
      }
    }
    disabled={false}
    onClick={[Function]}
  />
</div>
`;

exports[`components/<QuickReplies /> should render component with no data 1`] = `
<span
  className="red-text"
>
  Oops, something went wrong... :(
</span>
`;
